examples/nlp/question-answering.py

```diff
@@ -0,0 +1,422 @@
+        # span of text.
+        sample_index = sample_mapping[i]
+        answers = examples["answers"][sample_index]
+        # If no answers are given, set the cls_index as answer.
+        if len(answers["answer_start"]) == 0:
+            tokenized_examples["start_positions"].append(cls_index)
+            tokenized_examples["end_positions"].append(cls_index)
+        else:
+            # Start/end character index of the answer in the text.
+            start_char = answers["answer_start"][0]
+            end_char = start_char + len(answers["text"][0])
+
+            # Start token index of the current span in the text.
+            token_start_index = 0
+            while sequence_ids[token_start_index] != 1:
+                token_start_index += 1
+
+            # End token index of the current span in the text.
+            token_end_index = len(input_ids) - 1
+            while sequence_ids[token_end_index] != 1:
+                token_end_index -= 1
+
+            # Detect if the answer is out of the span (in which case this feature is labeled with the
+            # CLS index).
+            if not (
+                offsets[token_start_index][0] <= start_char
+                and offsets[token_end_index][1] >= end_char
+            ):
+                tokenized_examples["start_positions"].append(cls_index)
+                tokenized_examples["end_positions"].append(cls_index)
+            else:
+                # Otherwise move the token_start_index and token_end_index to the two ends of the
+                # answer.
+                # Note: we could go after the last offset if the answer is the last word (edge
+                # case).
+                while (
+                    token_start_index < len(offsets)
+                    and offsets[token_start_index][0] <= start_char
+                ):
+                    token_start_index += 1
+                tokenized_examples["start_positions"].append(token_start_index - 1)
+                while offsets[token_end_index][1] >= end_char:
+                    token_end_index -= 1
+                tokenized_examples["end_positions"].append(token_end_index + 1)
+
+    return tokenized_examples
+
+
+"""
+To apply this function on all the sentences (or pairs of sentences) in our dataset, we
+just use the `map` method of our `dataset` object, which will apply the function on all
+the elements of.
+
+We'll use `batched=True` to encode the texts in batches together. This is to leverage the
+full benefit of the fast tokenizer we loaded earlier, which will use multi-threading to
+treat the texts in a batch concurrently. We also use the `remove_columns` argument to
+remove the columns that existed before tokenization was applied - this ensures that the
+only features remaining are the ones we actually want to pass to our model.
+"""
+
+tokenized_datasets = datasets.map(
+    prepare_train_features, batched=True, remove_columns=datasets["train"].column_names
```

Comment:
Make sure to use `num_parallel_calls` (you can default to `tf.data.AUTOTUNE`)

Reply:
I used `num_proc` which is `datasets` equivalent of `num_parallel_calls`.